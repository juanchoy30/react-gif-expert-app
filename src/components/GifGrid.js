import React, { useState, useEffect } from 'react'
import { getGif } from '../helpers/getGifs';
import { GifGreedItem } from './GifGreedItem';

const GifGrid = ({ category }) => {

    const [images, setImages] = useState([]);

    useEffect( () => {
        getGif( category )
            .then( setImages );  // it is the same as .then( imgs => setImages( imgs ));
    }, [ category ]); // [] executes the function just once, if not the function could execute in an infinite cycle

    return (
        <>
            <h3>{ category }</h3>
            <div className="card-grid">
                {
                    images.map( img => (
                        <GifGreedItem 
                            key={ img.id }
                            { ...img }
                        />
                    ))
                }
            </div>
        </>

    )
}

export default GifGrid;