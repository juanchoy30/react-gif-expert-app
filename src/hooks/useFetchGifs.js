import { useEffect, useState } from "react";
import { getGifs } from '../helpers/getGifs';

export const useFetchGifs = ( category ) => {

    const [state, setState] = useState({
        data: [],
        loading: true
    });

    useEffect( () => {  // It cannot be async

        getGifs( category )
            .then( imgs => {
                setTimeout( () => {
                    setState({
                        data: imgs,
                        loading: false
                    });
                }, 2000);
            })

    }, [category]);

        /*
    useEffect( () => {
        getGif( category )
            .then( setImages );  // it is the same as .then( imgs => setImages( imgs ));
    }, [ category ]); // [] executes the function just once, if not the function could execute in an infinite cycle
    */

    return state;  // { data:[], loading: true };

}