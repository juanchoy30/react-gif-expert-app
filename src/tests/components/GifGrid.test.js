import React from 'react';
import '@testing-library/jest-dom';
import { shallow } from 'enzyme';

import GifGrid from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs')

describe('<GifGrid /> tests', () => {

    const category = 'X files'
    
    test('should render the component correctly', () => {

        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });

        const wrapper = shallow(<GifGrid category={ category }/>);
        expect( wrapper ).toMatchSnapshot();

    });

    test('should show images when loading useFetchGifs', () => {

        const gifs = [{
            id: 'ABC',
            url: 'https://localhost/any/thing.jpg',
            title: 'Anything'
        },
        {
            id: '123',
            url: 'https://localhost/any/thing.jpg',
            title: 'Anything'
        }];

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });
        
        const wrapper = shallow(<GifGrid category={ category }/>);

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('p').exists() ).toBe(false);
        expect( wrapper.find('GifGridItem').length ).toBe( gifs.length );

    })
    
    



})
