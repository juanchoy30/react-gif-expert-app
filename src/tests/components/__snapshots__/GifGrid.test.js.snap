// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GifGrid /> tests should render the component correctly 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    X files
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`<GifGrid /> tests should show images when loading useFetchGifs 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    X files
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="Anything"
      url="https://localhost/any/thing.jpg"
    />
    <GifGridItem
      id="123"
      key="123"
      title="Anything"
      url="https://localhost/any/thing.jpg"
    />
  </div>
</Fragment>
`;
